File: Dart/Structural/composite.dart
/// Class with commun operation
abstract interface class Element {
  double getPrice();
}

/// Class with the concrete implementation of comun
/// operation
class Box implements Element {
  Box({required this.elements});

  final List<Element> elements;

  /// In this case we get the price for each child
  @override
  double getPrice() => elements.fold(
      0, (previousValue, element) => previousValue + element.getPrice());
}

/// Class with the concrete implementation of comun
/// operation
class Product implements Element {
  Product({required double price}) : _price = price;

  final double _price;

  /// In this case simply return its price.
  @override
  double getPrice() => _price;
}

void main() {
  final product1 = Product(price: 15);
  final product2 = Product(price: 5);
  final product3 = Product(price: 5);
  final product4 = Product(price: 10);
  final product5 = Product(price: 5);

  final box1 = Box(elements: [product1, product2]);
  final box2 = Box(elements: [product3, product4]);

  final bigBox = Box(elements: [box1, box2, product5]);

  print('Elements in the box:');
  bigBox.elements.forEach((element) => print(element.runtimeType));
  print('Final price: ${bigBox.getPrice()}');

  /// Here is the result:
  /*
  Elements in the box:
  Box
  Box
  Product
  Final price: 40.0
  */
}

File: Swift/Structural/composite.swift
//
//  composite.swift
//
//
//

// Protocol with commun operation
protocol Element {
    func getPrice() -> Double
}

// Struct with the concrete implementation of comun
// operation
struct Box: Element {
    let elements: [any Element]

    init(elements: [any Element]) {
        self.elements = elements
    }

    // In this case we get the price for each child
    func getPrice() -> Double { elements.reduce(0) { $0 + $1.getPrice() } }
}

// Struct with the concrete implementation of comun
// operation
struct Product: Element {
    private let _price: Double

    init(price: Double) {
        self._price = price
    }

    // In this case simply return its price
    func getPrice() -> Double { _price }
}

// Demo
let product1 = Product(price: 15)
let product2 = Product(price: 5)
let product3 = Product(price: 5)
let product4 = Product(price: 10)
let product5 = Product(price: 5)

let box1 = Box(elements: [product1, product2])
let box2 = Box(elements: [product3, product4])

let bigBox = Box(elements: [box1, box2, product5])

print("Elements in the box:")
bigBox.elements.forEach { print(type(of: $0)) }
print("Final price: \(bigBox.getPrice())")

// Here is the result:
// Elements in the box:
// Box
// Box
// Product
// Final price: 40.0
